### PR TITLE
Stop reading from TCP connection if the client isn't reading from it.

### PR DESCRIPTION
**Pull request by Lukasa**
This patch addresses *literally exactly* the problem in https://twistedmatrix.com/trac/ticket/8868. This does not fully address this style of attack, only literally this instance.

### CODE REVIEW

File: src/twisted/internet/producer_helpers.py [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+
+        Stop streaming data to the consumer.
+        """
+        if self._finished:
+            return
+        self._finished = True
+        self._coopTask.stop()
+
+
+    # IPushProducer implementation:
```
Comment (hawkowl) [inline src/twisted/internet/producer_helpers.py]: Twistedchecker thinks this comment is also a blank line :(

File: src/twisted/internet/producer_helpers.py [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+    def pauseProducing(self):
```
Comment (hawkowl) [inline src/twisted/internet/producer_helpers.py]: Docstrings here (`@see IProducer.whatever` will suffice)

File: src/twisted/internet/test/test_tls.py [hunk 3]
```diff
@@ -340,16 +340,13 @@ class AbortSSLConnectionTests(ReactorBuilder, AbortConnectionMixin, ContextGener
 
     def buildReactor(self):
         reactor = ReactorBuilder.buildReactor(self)
-        try:
-            from twisted.protocols import tls
-        except ImportError:
-            return reactor
+        from twisted.internet import producer_helpers
 
         # Patch twisted.protocols.tls to use this reactor, until we get
```
Comment (hawkowl) [inline src/twisted/internet/test/test_tls.py:345]: Is this comment still relevant?

File: src/twisted/test/proto_helpers.py [hunk 4]
```diff
@@ -837,6 +838,46 @@ def connectUNIX(self, address, factory, timeout=30, checkPID=0):
...
+    A pull producer which writes 10 times only.
+    """
+
+    counter = 0
+    stopped = False
+
+    def __init__(self, consumer):
+        self.consumer = consumer
+        self.result = Deferred()
+
```
Comment (hawkowl) [inline src/twisted/test/proto_helpers.py:852]: Extra newline required here

File: src/twisted/internet/producer_helpers.py [hunk 5]
```diff
@@ -0,0 +1,113 @@
+# -*- test-case-name: twisted.test.test_producer_helpers -*-
```
Comment (hawkowl) [inline src/twisted/internet/producer_helpers.py]: Why is this a public module when it has no public items in it?

File: src/twisted/internet/producer_helpers.py [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+# Copyright (c) Twisted Matrix Laboratories.
```
Comment (hawkowl) [inline src/twisted/internet/producer_helpers.py]: This module also needs an `__all__`

**Review by hawkowl — DISMISSED** [review]
Please address the comments + the twistedchecker errors, then merge.

File: src/twisted/internet/test/test_tls.py (hunk 3 above)
Comment (Lukasa) [reply src/twisted/internet/test/test_tls.py:345]: I think so: 5206 was never fixed, but I think the breadcrumb is still useful.

**Comment by Lukasa** [comment]
Ok so txchecker genuinely wants it both ways. There is no way to have txchecker accept a comment before a method in Twisted: if there is one empty line then the comment it is too few blank lines, if there are two empty lines then there are too many blank lines.

I view this as a txchecker bug.

**Pull request merged**